node/network/availability-recovery/src/lib.rs

```diff
@@ -250,21 +262,48 @@ impl Interaction {
 					self.validators.len(),
 					self.received_chunks.values().map(|c| (&c.chunk[..], c.index as usize)),
 				) {
-					Ok(data) => FromInteraction::Concluded(self.candidate_hash.clone(), Ok(data)),
-					Err(_) => FromInteraction::Concluded(self.candidate_hash.clone(), Err(RecoveryError::Invalid)),
+					Ok(data) => {
+						if self.reconstructed_data_matches_root(&data) {
+							FromInteraction::Concluded(self.candidate_hash.clone(), Ok(data))
+						} else {
+							FromInteraction::Concluded(
+								self.candidate_hash.clone(),
+								Err(RecoveryError::Invalid),
+							)
+						}
+					}
+					Err(_) => FromInteraction::Concluded(
+						self.candidate_hash.clone(),
+						Err(RecoveryError::Invalid),
+					),
 				};
 
-				if let Err(e) = self.to_state.send(concluded).await {
-					tracing::warn!(
-						target: LOG_TARGET,
-						err = ?e,
-						"Failed to send Concluded result from interaction",
-					);
-				}
-				return;
+				self.to_state.send(concluded).await.map_err(error::Error::ClosedToState)?;
+				return Ok(());
 			}
 		}
 	}
+
+	fn reconstructed_data_matches_root(
```

Comment:
style: this could be a free function that accepts `n_validators` and `erasure_root` as parameters, to avoid reliance on `self`.